### ISSUE TITLE
When Threads>1 or ThreadPool>2 playbots are very prone to crashing  when there is more than one real player in the server

### ISSUE THREAD

**Issue by zbhcn** [liyunfan1223/mod-playerbots#715]
When configuring the option MapUpdate in the worldserver. conf file in the Azerothcore core When Threads>1 or ThreadPool>2, playbots are very prone to crashing when there is more than one real player in the server. I think it is most likely because playbots do not support multithreading

**Comment by zbhcn** [comment]
and When MapUpdate.Threads=1 or ThreadPool=2, no more crashes occur

**Comment by PaoloBenigni** [comment]
How many bots are you running? What are your system specs for the server? What is your memory usage? What about diff?

**Comment by zbhcn** [comment]
> How many bots are you running? What are your system specs for the server? What is your memory usage? What about diff?
> 
> Many people routinely run 3-4 mapupdate threads.

![5929c8f0c7a0e2aed6cf0968e02db207](https://github.com/user-attachments/assets/4b1535c0-fa8b-4615-821e-637235dbc7bc)

**Comment by zbhcn** [comment]
*** Hardware ***
Processor: AMD EPYC 74F3 24-Core Processor                
Number Of Processors: 16
Physical Memory: 16776756 KB (Available: 9117480 KB)
Commit Charge Limit: 19791412 KB

*** Operation System ***
Windows Server 2016 Datacenter (Version 10.0, Build 20348)

//=====================================================
Exception code: C0000420 出现声明故障。

Assertion message: 
#----------------------------------------------------------------------#
 
>> ABORTED

# Location 'G:\azerothcore-wotlk-Playerbot\src\server\apps\worldserver\Main.cpp:650'
# Function 'FreezeDetector::Handler'
# Message 'World Thread hangs for 240161 ms, forcing a crash!'

**Comment by PaoloBenigni** [comment]
Did you change `MaxCoreStuckTime = 0` in the worldserver.conf?

**Comment by zbhcn** [comment]
> Did you change `MaxCoreStuckTime = 0` in the worldserver.conf?

yes, i set MaxCoreStuckTime = 240

**Comment by PaoloBenigni** [comment]
Was there a specific reason? If the diff is typically less than 100 ms, and only sometimes you get a spike then I would just leave it off.

**Comment by PaoloBenigni** [comment]
I would turn that setting off then unless you see issues. It should eliminate the error.

**Comment by noisiver** [comment]
There is no real reason to change `ThreadPool`. Only `MapUpdate.Threads` make a noticeable difference.

This is a Windows issue as far as I can tell. I can run the server on Linux for weeks without any issues but on Windows it takes very little time before it locks up. The settings across both systems are identical. I don't mess with `ThreadPool` but I set `MapUpdate.Threads` to 4. I haven't tried leaving `MapUpdate.Threads` on 1 though.

**Comment by zbhcn** [comment]
> There is no real reason to change `ThreadPool`. Only `MapUpdate.Threads` makes a noticeable difference.
> 
> This is a Windows issue as far as I can tell. I can run the server on Linux for weeks without any issues but on Windows it takes very little time before it locks up. The settings across both systems are identical. I don't mess with `ThreadPool` but I set `MapUpdate.Threads` to 4. I haven't tried leaving `MapUpdate.Threads` on 1 though.
> 
> Edit: I should mention this is not likely to be because the server is overloaded. The only time I have random bots logged in is if I want to run some battlegrounds or arenas and that rarely happens. It froze up last night when I was logged in with 24 altbots (plus my own character) and zero rndbots.
> 
> It's not because of multiple players logged in. I was logged in for maybe 10 minutes before it locked up the last time, and no one but me has access to the server.
> 
> I don't know why it happens but it feels kind of crucial to fix.

thank you for your reply, there is no altbot in my server except 260 rndbots, They play random dungeons and battlefields and arenas with more than 20 real players, and in the raid, I only allow players to team up with healer bots, and up to a maximum of 6. At the moment, it is generally in line with my ideal gaming environment.

**Comment by noisiver** [comment]
I'm marking this as Medium until further notice. I don't know if it qualifies as Urgent but it sure is a bad one. I have no idea how you would go about figuring out why it's happening though.

I hope more people who are experiencing this can chime in. It would be good to figure out if this is in fact a Windows thing or if someone is experiencing it on Linux as well.

**Comment by zbhcn** [comment]
I just checked the server logs and found another issue of downtime. My setting is MapUpdate Threads = 4
[unknown_worldserver.exe_[17-11_5-11-25].txt](https://github.com/user-attachments/files/17788222/unknown_worldserver.exe_.17-11_5-11-25.txt)

**Comment by zbhcn** [comment]
Confirm again when MapUpdate When Threads>1, downtime issues are very frequent.Especially when robots use LFG

**Comment by noisiver** [comment]
If you keep `MapUpdate.Threads` set to 1, does it freeze at all?

When it froze for me the other day I was just flying around Icecrown looking for quest objects I needed. I didn't even land or aggro anything, I was just flying around.

**Comment by zbhcn** [comment]
When MapUpdate.Threads=1 and ThreadPool=2, no more crashes occur

**Comment by zbhcn** [comment]
> If you keep `MapUpdate.Threads` set to 1, does it freeze at all?
> 
> When it froze for me the other day I was just flying around Icecrown looking for quest objects I needed. I didn't even land or aggro anything, I was just flying around.


 When MapUpdate.Threads=1 and ThreadPool=2, no more crashes occur. Your server should only have you as a real player, right?

**Comment by noisiver** [comment]
I am the only player on my realms, yep. There are never any other players.

**Comment by zbhcn** [comment]
> I am the only player on my realms, yep. There are never any other players.

So you'll never run into a problem that is only available on multiplayer servers

**Comment by noisiver** [comment]
No, but I have run into this so it can't be caused by you having multiple players online.

**Comment by zbhcn** [comment]
> 不，但是我遇到过这种情况，所以这不是由多个玩家在线造成的。

I've found that it's easy to crash when multiple players are using lfg at the same time, and mapupdate.thread >1

**Comment by zbhcn** [comment]
> No, but I have run into this so it can't be caused by you having multiple players online.

When MapUpdate.Threads=1, no more crashes occur , So I think it must be mapupdate.thread

**Comment by maikale** [comment]
I'm having a lag problem too 250 bots after 10 15 minutes it becomes scary where I'm messing up
*** Hardware ***
Processor: Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz
Number Of Processors: 24
Physical Memory: 65457 MB
*** Operation System ***
Windows 10 Professional

**Comment by noisiver** [comment]
> I'm having a lag problem too 250 bots after 10 15 minutes it becomes scary where I'm messing up *** Hardware *** Processor: Intel(R) Xeon(R) CPU E5-2690 v3 @ 2.60GHz Number Of Processors: 24 Physical Memory: 65457 MB *** Operation System *** Windows 10 Professional

I'm not sure I understand what your issue is. We're talking about the fact that the server freezes and is no longer responding. The only way out of it is to kill the server.

**Comment by noisiver** [comment]
With 1 map update thread I haven't seen it freeze yet. I'm going to give it a lot more time to make sure but so far it looks like increasing the thread count is somehow causing it.
![image](https://github.com/user-attachments/assets/d685dffb-b4bb-49a4-bb88-ef95bd9d9725)

**Comment by maikale** [comment]
> > И аз имам проблем със забавяне 250 бота след 10 15 минути става страшно къде бъркам *** Хардуер *** Процесор: Intel(R) Xeon(R) CPU E5-2690 v3 @ 2,60GHz Брой Процесори: 24 Физическа памет: 65457 MB *** Операционна система *** Windows 10 Professional
> 
> Не съм сигурен, че разбирам какъв е проблемът ви. Говорим за факта, че сървърът замръзва и вече не отговаря. Единственият изход от това е да убиете сървъра.

For me it becomes high lag unplayable

**Comment by PaoloBenigni** [comment]
Share how many bots you have, if any players, smart scales we on or off and your hardware.

**Comment by maikale** [comment]
![image](https://github.com/user-attachments/assets/0409bb9c-6785-4896-8f30-88957b4346c9)

**Comment by maikale** [comment]
![WoWScrnShot_111724_185329](https://github.com/user-attachments/assets/a9bf93c5-ed59-4eef-8174-b0b8b3f0eea2)

**Comment by maikale** [comment]
250 bots + me

**Comment by hermensbas** [comment]
I am not suprised as in the playerbot does not contain alot of threadsafe code checks, not entirely sure whether thats part of the problem though.

**Comment by noisiver** [comment]
I'm still convinced this is only an issue on Windows, but increasing the thread count will absolutely cause it (for some or most people at least). I have been running the server on Windows for 2 days now and no crashes with `MapUpdate.Threads` set to 1. I never had issues with this on Linux.

**Comment by hermensbas** [comment]
> I'm still convinced this is only an issue on Windows, but increasing the thread count will absolutely cause it (for some or most people at least). I have been running the server on Windows for 2 days now and no crashes with `MapUpdate.Threads` set to 1. I never had issues with this on Linux.

Could be, i never run the game in windows ever :P

**Comment by noisiver** [comment]
I have been running a server on Windows for 4 days now, map update threads set to 4, and I've tried with and without random bots and it hasn't frozen yet. I don't know if a core update solved it on my end or what but right now I can't seem to replicate this anymore.

**Comment by zbhcn** [comment]
> I have been running a server on Windows for 4 days now, map update threads set to 4, and I've tried with and without random bots and it hasn't frozen yet. I don't know if a core update solved it on my end or what but right now I can't seem to replicate this anymore.

I tested on a Debian system yesterday, and found that when MapUpdate.Threads > 1, the worldserver crashes after running for a few minutes to several minutes. Of course, there are multiple players on my server, and it does not crash when there is only one player.

**Comment by davvvve1** [comment]
I have freeze when im doing dungeons - Just when i spawn back server freeze- 
I turn all settings to 1 and kept threadpool to 8 . Now it happes rarely . Gonna try some more . Im on Debian 12 . I have 3k bots .

**Comment by davvvve1** [comment]
> I have been running a server on Windows for 4 days now, map update threads set to 4, and I've tried with and without random bots and it hasn't frozen yet. I don't know if a core update solved it on my end or what but right now I can't seem to replicate this anymore.

Try to spam dungeons . After 2-3 the server freeze

**Comment by noisiver** [comment]
I constantly spam dungeons.

**Comment by zbhcn** [comment]
> I am not suprised as in the playerbot does not contain alot of threadsafe code checks, not entirely sure whether thats part of the problem though.
> 
> @zbhcn indicates this happens in certain situations especially when multiple real-players online next to the bot. Lets keep the bug open. Without multiple crashlogs or crashlogs conclusion its gonna be tricky to analyse though.

You're absolutely right. Due to the large number of bots, the bot code should ideally run in multiple separate threads to improve overall core efficiency. However, this functionality has not been implemented yet, and the bot code is still running in the main thread. By the way, as far as I know, the Eluna module is also single-threaded, so this crash issue could potentially be caused by Eluna's code, especially when there are multiple real players involved.

**Comment by hermensbas** [comment]
Well, i kinda assume when ppl make bugs they test the bugs in a clean builds without additional mods. My test/dev build is always clean and default. The core might support a plugin model in shape of modules but the construction can still introduce a lot of side effects.

**Comment by zbhcn** [comment]
> [@zbhcn](https://github.com/zbhcn), do you still have this issue with the latest revisions? Some rework was done on the mapupdater class meanwhile. Also, did/does it happen on a build without the playerbots add-on, so just AzerothCore and nothing else?

yes ， This issue did not occur in a build without the Playerbots add-on , it still have this issue with the latest revisions
